src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Linux.cs

```diff
@@ -0,0 +1,1200 @@
+                ThreadPool.QueueUserWorkItem(_ => callback(acceptedFd, socketAddress, socketAddressLen, errorCode), null);
+                return errorCode;
+            }
+
+            var operation = new AcceptOperation {
+                Callback = callback,
+                SocketAddress = socketAddress,
+                SocketAddressLen = socketAddressLen
+            };
+
+            bool isStopped;
+            while (!TryBeginOperation(ref _acceptOrConnectQueue, operation, out isStopped))
+            {
+                if (isStopped)
+                {
+                    // TODO: is this error reasonable for a closed socket? Check with Winsock.
+                    operation.ErrorCode = SocketError.Shutdown;
+                    QueueCompletion(operation);
+                    return SocketError.Shutdown;
+                }
+
+                if (TryCompleteAccept(_fileDescriptor, operation))
+                {
+                    QueueCompletion(operation);
+                    break;
+                }
+            }
+            return SocketError.IOPending;
+        }
+
+        private static bool TryCompleteAccept(int fileDescriptor, AcceptOperation operation)
+        {
+            return TryCompleteAccept(fileDescriptor, operation.SocketAddress, ref operation.SocketAddressLen, out operation.AcceptedFileDescriptor, out operation.ErrorCode);
+        }
+
+        private static unsafe bool TryCompleteAccept(int fileDescriptor, byte[] socketAddress, ref int socketAddressLen, out int acceptedFd, out SocketError errorCode)
+        {
+            int fd;
+            uint sockAddrLen = (uint)socketAddressLen;
+            fixed (byte* rawSocketAddress = socketAddress)
+            {
+                fd = Interop.libc.accept(fileDescriptor, (Interop.libc.sockaddr*)rawSocketAddress, &sockAddrLen);
+            }
+
+            if (fd != -1)
+            {
+                socketAddressLen = (int)sockAddrLen;
+                errorCode = SocketError.Success;
+                acceptedFd = fd;
+                return true;
+            }
+            acceptedFd = -1;
+
+            Interop.Error errno = Interop.Sys.GetLastError();
+            if (errno != Interop.Error.EAGAIN && errno != Interop.Error.EWOULDBLOCK)
+            {
+                errorCode = SocketPal.GetSocketErrorForErrorCode(errno);
+                return true;
+            }
+
+            errorCode = SocketError.Success;
+            return false;
+        }
+
+        public SocketError ConnectAsync(byte[] socketAddress, int socketAddressLen, Action<SocketError> callback)
+        {
+            Debug.Assert(socketAddress != null);
+            Debug.Assert(socketAddressLen > 0);
+            Debug.Assert(callback != null);
+
+            SocketError errorCode;
+            if (TryCompleteConnect(_fileDescriptor, socketAddress, socketAddressLen, out errorCode))
+            {
+                ThreadPool.QueueUserWorkItem(_ => callback(errorCode), null);
```

Comment:
I won't comment on these closures further; at some point it'd be good to do a pass through to fix up such low-hanging fruit.  But I think you said many of these synchronously completing cases are likely to go away anyway, such that we don't invoke the callback asynchronously (great!), in which case the problem will mostly solve itself.